core/executor/src/wasm_runtimes_cache.rs

```diff
@@ -0,0 +1,328 @@
+	/// The snapshot of the instance's state taken just after the instantiation.
+	state_snapshot: StateSnapshot,
+}
+
+impl CachedRuntime {
+	/// Perform an operation with the clean version of the runtime wasm instance.
+	pub fn with<R, F>(&self, f: F) -> R
+	where
+		F: FnOnce(&WasmModuleInstanceRef) -> R,
+	{
+		self.state_snapshot.apply(&self.instance).expect(
+			"applying the snapshot can only fail if the passed instance is different
+			from the one that was used for creation of the snapshot;
+			we use the snapshot that is directly associated with the instance;
+			thus the snapshot was created using the instance;
+			qed",
+		);
+		f(&self.instance)
+	}
+
+	/// Returns the version of this cached runtime.
+	///
+	/// Returns `None` if the runtime doesn't provide the information or there was an error
+	/// while fetching it.
+	pub fn version(&self) -> Option<RuntimeVersion> {
+		self.version.clone()
+	}
+}
+
+/// A state snapshot of an instance taken just after instantiation.
+///
+/// It is used for restoring the state of the module after execution.
+#[derive(Clone)]
+struct StateSnapshot {
+	/// The offset and the content of the memory segments that should be used to restore the snapshot
+	data_segments: Vec<(u32, Vec<u8>)>,
+	/// The list of all global mutable variables of the module in their sequential order.
+	global_mut_values: Vec<RuntimeValue>,
+	heap_pages: u32,
+}
+
+impl StateSnapshot {
+	// Returns `None` if instance is not valid.
+	fn take(
+		module_instance: &WasmModuleInstanceRef,
+		data_segments: Vec<DataSegment>,
+		heap_pages: u32,
+	) -> Option<Self> {
+		let mut prepared_segments = Vec::with_capacity(data_segments.len());
+		for mut segment in data_segments {
+			// Just replace contents of the segment since the segments will be discarded later
+			// anyway.
+			let contents = mem::replace(segment.value_mut(), vec![]);
+
+			let init_expr = segment.offset().code();
+			// [op, End]
+			if init_expr.len() != 2 {
+				return None;
+			}
+			let offset = match init_expr[0] {
+				Instruction::I32Const(v) => v as u32,
+				Instruction::GetGlobal(idx) => {
+					let global_val = module_instance.globals().get(idx as usize)?.get();
+					match global_val {
+						RuntimeValue::I32(v) => v as u32,
+						_ => return None,
+					}
+				}
+				_ => return None,
+			};
+			prepared_segments.push((offset, contents))
+		}
+
+		// Collect all values of mutable globals.
+		let global_mut_values = module_instance
+			.globals()
+			.iter()
+			.filter(|g| g.is_mutable())
+			.map(|g| g.get())
+			.collect();
+
+		Some(Self {
+			data_segments: prepared_segments,
+			global_mut_values,
+			heap_pages,
+		})
+	}
+
+	/// Reset the runtime instance to the initial version by restoring
+	/// the preserved memory and globals.
+	///
+	/// Returns `Err` if applying the snapshot is failed.
+	fn apply(&self, instance: &WasmModuleInstanceRef) -> Result<(), CacheError> {
+		let memory = instance
+			.export_by_name("memory")
```

Comment:
Theoretically `export_by_name` is more proper since there is no way to ensure that the right memory will be at index 0. Practically, however, since at the moment at most one memory is allowed, we could use it (Although we would need to export `memory_by_index`, it is private right now).

However, IMO I think we should stick to names. I had some ideas about extracting a cached version of runtime instance where things like `MemoryRef` are preloaded. According to my profiling this should give noticable improvements.